design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+
+* as many prefixes as chunks, for example by selecting `s3://some-bucket/some-prefix/c/0/0/1` in the second example above
+* too large of an extra string in each chunk, for example, if we select `s3://some-bucket` as the prefix each chunk needs to add something like `some-prefix/c/0/0`.
+
+### Should prefixes be per snapshot or global?
+
+The advantage of per snapshot is that we don't need to load/parse all the prefixes if the current snapshot uses only a few. A case for this would be a virtual dataset that has been materialized in a newer version. If prefixes are per-snapshot we wouldn't need to load any prefixes.
+
+On the other hand, requirement 4 asks for the full list of "dependencies" for the dataset, and that means for all versions, which requires a global list of prefixes. Without a global list, we would need to fetch the list of prefixes of each snapshot.
+
+### Who creates/decides the prefixes
+
+Obviously, generating the list of prefixes given the list of virtual chunks is doable automatically. But it's a tricky optimization problem. One strategy could be to generate a trie with all chunk urls, and then use some heuristic to select prefixes using heavily populated nodes. It's tricky and potentially computationally expensive for large repos.
+
+An alternative is to let the user dictate the prefixes. This way, they can optimize them knowing beforehand what virtual chunks they will add. If they don't optimize, they get large manifests without prefix extraction. We still need to satisfy requirements 2 and 4, which would require automatic extraction of at least the bucket and platform.
+
+Additionally, the optimization problem can be treated as an extra process (similar to compaction or garbage collection), and ran offline on demand.
+
+### Templates instead of prefixes
+
+Using templates would be more efficient, example:
+
+* `s3://some-bucket/some-prefix/1-analysis-output-for-model-abc.nc`
+* `s3://some-bucket/some-prefix/2-analysis-output-for-model-abc.nc`
+
+If we use the template `s3://some-bucket/some-prefix/{}-analysis-output-for-model-abc.nc` each chunk could have a couple of extra characters only, filling the placeholders in the template.
+
+### Supporting multiple clouds
+
+When the user adds a virtual ref, they need to indicate the details of the platform to retrieve it. These include what cloud it is (so we know how to set up credentials), region, endpoint url, port and other settings. It looks a lot like our `S3Config` type (or an expanded version of that).
+
+### Supporting multiple credentials
+
+Credentials are a read time concern. When the user writes the virtual chunk, they don't necessarily know what credentials set will be used to read it.
+
+Credentials need to be specified later, when the chunks are going to be read. For example, by indicating in runtime configuration a set of credentials for each virtual chunk bucket/prefix/template.
```

Comment:
I feel we need to do more work to figure out how we will be mapping Prefixes / Dependencies / Locations / VirtualChunkContainers (whatever we end up calling them) to credential sets.